### PR TITLE
Stabilize auth refresh tests

### PR DESCRIPTION
**Pull request by eiri**
## PR summary

Auth test "Test authentication refresh" occasionally fails due to async nature of the refresh mechanism. We are running three refresh clients in parallel with an assumption that the refresh process is slow enough for all of them to spawn before it finishes.

This proven to be unreliable, leading to a flaky test, so this change makes sure that we are testing only _first_ of the refresh clients to confirm that it is starting the refresh process asynchronously and then observing req counter to make sure we are calling our test server only once.

Fixes: #99

**Note: An existing issue is [required](https://github.com/IBM/cloudant-go-sdk/blob/master/CONTRIBUTING.md#PRs) before opening a PR.**

## PR Checklist

Please make sure that your PR fulfills the following requirements:

- [x] The commit message follows the
[Angular Commit Message Guidelines](https://github.com/angular/angular/blob/master/CONTRIBUTING.md#-commit-message-guidelines).
- [x] Tests for the changes have been added (for bug fixes / features)
- [ ] Docs have been added / updated (for bug fixes / features)

## PR Type  

- [x] Bugfix
- [ ] Feature
- [ ] Code style update (formatting, local variables)
- [ ] Refactoring (no functional changes, no api changes)
- [ ] New tests
- [ ] Build/CI related changes
- [ ] Documentation content changes
- [ ] Other (please describe)

## What is the current behavior?
Occasional failure of "Test authentication refresh" test when auth session updated to fast from observer perspective

## What is the new behavior?

More stable behaviour of "Test authentication refresh" test

## Does this PR introduce a breaking change?

- [ ] Yes
- [x] No

### CODE REVIEW

File: auth/couchdb_session_authenticator_test.go [hunk 1]
```diff
@@ -232,25 +233,36 @@ var _ = Describe("Authenticator Unit Tests", func() {
...
+						Expect(cookie.Value).To(Equal("fakefake-1"))
+					}
+				}(i)
 			}
 			wg.Wait()
 
-			// give refresher goroutine time to finish
-			time.Sleep(1 * time.Second)
+			// wait until server counter increased by refresher
+			Eventually(func() int {
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:252]: Nice find of this `Eventually` function! 👍

File: auth/couchdb_session_authenticator_test.go [hunk 2]
```diff
@@ -232,25 +233,36 @@ var _ = Describe("Authenticator Unit Tests", func() {
...
+			}).Should(Equal(2))
 
 			newCookie, err := auth.getCookie()
 			Expect(err).To(BeNil())
 			Expect(newCookie).ToNot(Equal(oldCookie))
 			Expect(newCookie.Value).To(Equal("fakefake-2"))
 
+			// wait a bit to confirm that we haven't had hits
+			// from some late refresh process
+			time.Sleep(100 * time.Millisecond)
```
Comment (vmatyusGitHub) [inline auth/couchdb_session_authenticator_test.go:263]: Nit: alternatively ca be used here the pair of the Eventually function: `Consistently`

**Review by vmatyusGitHub — COMMENTED** [review]
This solution must be much more stable.

**Review by vmatyusGitHub — APPROVED** [review]
LGTM! 👍

File: auth/couchdb_session_authenticator_test.go [hunk 3]
```diff
@@ -232,25 +233,36 @@ var _ = Describe("Authenticator Unit Tests", func() {
...
-			// give refresher goroutine time to finish
-			time.Sleep(1 * time.Second)
+			// wait until server counter increased by refresher
+			Eventually(func() int {
+				return int(atomic.LoadInt32(&callNumber))
+			}).Should(Equal(2))
 
 			newCookie, err := auth.getCookie()
 			Expect(err).To(BeNil())
 			Expect(newCookie).ToNot(Equal(oldCookie))
```
Comment (ricellis) [inline auth/couchdb_session_authenticator_test.go:258]: I saw this assertion fail when running locally with this branch - I think that means the removal of the 1 second wait has opened a new window where the session call increment has happened, but the parsing of the response cookie and passing of it to the refresh channel has not yet completed.

I suggest we just put this subsequent `getCookie()` assertion in an `Eventually` with a timeout, so we assert immediately if the refresh has fully completed or poll a few times if it hasn't.

**Review by ricellis — CHANGES_REQUESTED** [review]

File: auth/couchdb_session_authenticator_test.go (hunk 3 above)
Comment (vmatyusGitHub) [reply auth/couchdb_session_authenticator_test.go]: How you execute the tests that you find these cases? Do you use a bash loop or something else?

Comment (ricellis) [reply auth/couchdb_session_authenticator_test.go]: Sure a loop, but it wasn't really necessary, timing is such on my machine that I never needed to get past 2 invocations to see a failure.

**Review by ricellis — APPROVED** [review]

File: auth/couchdb_session_authenticator_test.go (hunk 3 above)
Comment (vmatyusGitHub) [reply auth/couchdb_session_authenticator_test.go]: With this bash script I could not catch the problem:
```bash
#!/usr/bin/env bash

counter=1
while [ $counter -le 10 ]
do
  go clean -testcache
  go test ./auth
  ((counter++))
done
echo All done
```
What is yours?

Comment (eiri) [reply auth/couchdb_session_authenticator_test.go]: fwiw I couldn't reproduce it even in ~10min loop, so we are lucky @ricellis has such a _special_ MBP, close by efficiency to a resource limited Jenkins 😛

yet this case should be fixed now with me calling it in `Eventually` with the latest change I pushed.

I'm going to merge this patch. It's not ideal, but hopefully will be stable enough to not cause the issues in CI

Comment (vmatyusGitHub) [reply auth/couchdb_session_authenticator_test.go]: 🤞

**Pull request merged**